### ISSUE TITLE
Add references to Node & Edge

### ISSUE THREAD

**Issue by keyserj** [amelioro/ameliorate#114]
**Is your feature request related to a problem? Please describe.**
This is a refactor. Right now, if you have a node, you also need the diagram it's in before you can know anything about the node's related edges & neighbors. Most operations dealing with a node need to know some of this information, and passing the diagram around everywhere becomes cumbersome (and not-performant, since edges/neighbors need to be re-found).

**Describe the solution you'd like**
Node should have references to:
- parents
- children
- neighbors
- edges
- components

and Edge should have references to:
- parent
- child
- nodes

and both should have references to:
- diagram
- claims

and claims should have references to:
- critiquedArguable

**Describe alternatives you've considered**

**Additional context**

**Technical ideas**
Need to ensure these references are properly kept up-to-date. Also not sure how to populate these values on page-load; hopefully there's an on-load function for Zustand's persist middleware?

**Comment by keyserj** [comment]
See https://ameliorate.app/keyserj/backend-design for an example of a topic that is extremely slow to interact with (modify text, select a node) due to its size.

A lot of the lag is caused by hooks that do a lot of iteration, and are used by many nodes/edges.

<details><summary>Here are performance notes that I took - it seems like a lot of performance could be saved by this refactor.</summary>

Hooks performance eval for page load of backend tech topic:

note: these all seem to be firing 4 times per page load as well, could investigate that too.

useIsAnyGraphPartSelected: 252ms (2520 x 0.1ms)
-> could just track `isAnySelected` in the store

useIsNeighborSelected: 195ms (1304 x 0.15ms)
-> node.neighbors would make this way faster
-> if we set node.data.neighbors, I think we can avoid a hook entirely

useIsEdgeSelected: 130ms (1304 x 0.1ms)
-> node.edges would make this way faster
-> if we set node.data.edges, I think we can avoid a hook entirely

useNeighbors: 260ms (2608 x 0.1ms)
-> same as above, if we add node.data.parents, node.data.children

useNodes: 7ms (7 x 0.1ms)
-> this is only used for the topic pane, so no big deal

useNodeChildren: 130ms (1304 x 0.1ms)
-> same as above if node.data.children is added
-> this is only used for the criteria table indicator!

useNode: 130ms (1304 x 0.1ms)
-> same as above if node.data.children is added
-> this is only used for the criteria table indicator!
-> maybe not entirely replaceable but some usages can be replaced by just passing `node` instead of `nodeId` into the component

useIsImplied: 1806ms (1204 x 1.5ms)
-> curiously, this only ran once, unlike the above hooks that ran ~4 times after pageload
-> could store edge.data.isImplied; might be annoying because every view change would have to recalc, but might be worth

useIsNodeSelected: 120ms (1204 x 0.1ms)
-> curiously, this only ran once, unlike the above hooks that ran ~4 times after pageload
-> could store edge.data.nodes to avoid needing a hook

useFilteredDiagram: 12ms (8 x 1.5ms)

------
render times

ScoreEdge: 3ms per
-> seems like the edge hooks are causing this

FlowNode: generally < 1ms per, but there are a lot of components per FlowNode

<details/>
